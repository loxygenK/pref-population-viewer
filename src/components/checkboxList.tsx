import React from "react";
import { ValueWithID } from "~/types/valueWithId";
import styles from "./checkboxList.module.scss";
import { ToggleButton } from "./toggleButton";

export interface CheckBoxListProps {
  values: Array<ValueWithID<string>>;
  checkedIDs: Array<string>;
  onChange: (checkedIDs: Array<string>) => void;
}
export const CheckBoxList: React.FC<CheckBoxListProps> = ({
  values,
  checkedIDs,
  onChange,
}) => {
  const [shown, setShown] = React.useState(true);

  const onShowToggleButtonClicked = (newShownState: boolean) => {
    setShown(newShownState);
  };

  const onCheckboxChanged =
    (id: string) => (e: React.ChangeEvent<HTMLInputElement>) => {
      let newCheckedIDs = [...checkedIDs];
      if (e.target.checked) {
        newCheckedIDs.push(id);
      } else {
        newCheckedIDs = newCheckedIDs.filter((e) => e !== id);
      }

      onChange(newCheckedIDs);
    };

  return (
    <div className={styles.checkbox_list_wrapper}>
      <ToggleButton
        className={styles.show_toggle_button}
        captionWhenPressed="折りたたむ"
        captionWhenUnpressed="表示する"
        pressed={shown}
        onChange={onShowToggleButtonClicked}
      />
      {shown && (
        <ul className={styles.checkbox_list}>
          {values.map((v) => (
            <li key={v.id}>
              <input
                id={`check-${v.id}`}
                type="checkbox"
                onChange={onCheckboxChanged(v.id)}
                checked={checkedIDs.indexOf(v.id) !== -1}
              />
              <label htmlFor={`check-${v.id}`}>{v.value}</label>
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};
