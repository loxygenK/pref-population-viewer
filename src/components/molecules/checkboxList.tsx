import React from "react";
import { ValueWithID } from "~/types/valueWithId";
import styles from "./checkboxList.module.scss";
import { ToggleButton } from "../atom/toggleButton";
import { Checkbox } from "../atom/checkbox";

export interface CheckBoxListProps {
  values: Array<ValueWithID<string>>;
  checkedIDs: Array<string>;
  onChange: (checkedIDs: Array<string>) => void;
}
export const CheckBoxList: React.FC<CheckBoxListProps> = ({
  values,
  checkedIDs,
  onChange,
}) => {
  const [shown, setShown] = React.useState(true);

  const onShowToggleButtonClicked = (newShownState: boolean) => {
    setShown(newShownState);
  };

  const onCheckboxChanged = (id: string) => (becomeChecked: boolean) => {
    let newCheckedIDs = [...checkedIDs];
    if (becomeChecked) {
      newCheckedIDs.push(id);
    } else {
      newCheckedIDs = newCheckedIDs.filter((e) => e !== id);
    }

    onChange(newCheckedIDs);
  };

  return (
    <div className={styles.checkbox_list_wrapper}>
      <ToggleButton
        className={styles.show_toggle_button}
        captionWhenPressed="折りたたむ"
        captionWhenUnpressed="表示する"
        pressed={shown}
        onChange={onShowToggleButtonClicked}
      />
      <ul className={`${styles.checkbox_list} ${shown ? "" : styles.collapse}`}>
        {values.map((v) => (
          <li key={v.id}>
            <Checkbox
              id={`check-${v.id}`}
              onChange={onCheckboxChanged(v.id)}
              checked={checkedIDs.indexOf(v.id) !== -1}
              label={v.value}
            />
          </li>
        ))}
      </ul>
    </div>
  );
};
