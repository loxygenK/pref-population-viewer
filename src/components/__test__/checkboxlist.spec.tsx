import renderer from "react-test-renderer";
import { CheckBoxList } from "../checkboxList";

describe("Checkbox list component", () => {
  it("should render correctly", () => {
    const tree = renderer
      .create(
        <CheckBoxList
          values={Array.from({ length: 10 }).map((_, i) => ({
            id: i.toString(),
            value: `ID ${i}`,
          }))}
          checkedIDs={Array.from({ length: 10 }).map((_, i) =>
            (i * 2).toString()
          )}
          onChange={() => {}}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
