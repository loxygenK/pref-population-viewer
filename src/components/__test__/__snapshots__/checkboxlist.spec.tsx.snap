// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox list component should render correctly 1`] = `
<div
  className="checkbox_list_wrapper"
>
  <button
    className="toggle_button pressed show_toggle_button"
    onClick={[Function]}
  >
    折りたたむ
  </button>
  <ul
    className="checkbox_list "
  >
    <li>
      <input
        checked={true}
        id="check-0"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-0"
      >
        ID 0
      </label>
    </li>
    <li>
      <input
        checked={false}
        id="check-1"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-1"
      >
        ID 1
      </label>
    </li>
    <li>
      <input
        checked={true}
        id="check-2"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-2"
      >
        ID 2
      </label>
    </li>
    <li>
      <input
        checked={false}
        id="check-3"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-3"
      >
        ID 3
      </label>
    </li>
    <li>
      <input
        checked={true}
        id="check-4"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-4"
      >
        ID 4
      </label>
    </li>
    <li>
      <input
        checked={false}
        id="check-5"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-5"
      >
        ID 5
      </label>
    </li>
    <li>
      <input
        checked={true}
        id="check-6"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-6"
      >
        ID 6
      </label>
    </li>
    <li>
      <input
        checked={false}
        id="check-7"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-7"
      >
        ID 7
      </label>
    </li>
    <li>
      <input
        checked={true}
        id="check-8"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-8"
      >
        ID 8
      </label>
    </li>
    <li>
      <input
        checked={false}
        id="check-9"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        htmlFor="check-9"
      >
        ID 9
      </label>
    </li>
  </ul>
</div>
`;
